plat: Rework paged memory init

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This series reworks the initialization of paged memory to provide a more flexible implementation that is capable of handling regions beyond the limits defined in the boot page-tables. The motivation for this change is to allow mapping device regions that are unknown at compile-time, such as _Unprotected IPA Alias_ regions of Arm CCA Realms [1][2], the address of which depends on the executing platform.

This rework has side-effects on the way device MMIO regions are mapped. To address these requirements, this PR performs the following changes:
* Introduce early init boot stage in `kvm/arm64`.
* Add `early_init` to drivers that use MMIO, and have these drivers register bootinfo mrds for these regions during `early_init`.
* Add runtime mapping of MMIO regions to the initialization of relevant drivers.

All the above are described in further detail in the sections that follow.

Additional fixes include:
* lib/ukbus: Use the correct init class and priority.
* plat/drivers/uktty: Clean up driver init (`ns16550` and `pl011`).
* plat/drivers/uktty: Move driver init to `uk_inttab` (`ns16550` and `pl011`).
* plat/kvm/arm64: Remove console registration from platform setup.

#### Rework of paged memory init

Under the new scheme, bootinfo is simplified to only contain valid memory regions. Upon paged memory initialization these regions are either mapped with the protections defined in their mrd, or are already mapped, in which case only their protections are updated. An exception to this is mrds of the `UKPLAT_MEMRT_DEVICE` type, that are ignored to allow MMIO regions mapped during early device initialization to remain mapped.
    
Any memory not defined in bootinfo is unmapped. The theoretical bounds used as a reference are platform-specific and are defined in a newly introduced `platform.h` that must be provided by all platforms.

#### Early device initialization

`early_init` is invoked as the last call of early boot code, as soon as the stack and MMU are enabled.
    
Early devices are initialized before bootinfo setup, as the latter depends on UART for debug messages.

Early devices can use the APIs provided by libfdt / ACPI to obtain any information required from the firmware, such as device regions and the kernel command line.
    
Drivers that register with `early_init` are updated to append mrds of their MMIO regions to `bootinfo`. That is required so that their regions are not unmapped by paged memory init. The mrds added during `early_init` to bootinfo must use the newly introduced `UKPLAT_MEMRT_DEVICE` type.

Be aware that early drivers should not call `ukplat_bootinfo_coalesce`, as mrd coalescing is performed once at the end of `bootinfo_setup`, which is now performed as to the last step of `early_init`.

Notice that:
* `early_init` is currently introduced in `kvm/arm64`, as it is the only platform that involves early initialization of drivers that use MMIO. 
* `early_init` is currently implemented as a simple placeholder from where drivers can add manual calls of their early_init functions. This should be replaced with an early inittab as soon as bootinfo is moved out of `plat`.

#### Late (i.e. non-early) device initialization

Given that all memory that is not backed by an mrd is unmapped during paged memory init, devices that don't register mrds during early init are now required to map their regions during probing. To address this requirement this PR updates relevant drivers to map their MMIO regions dynamically.

#### References

[1] Realm Management Monitor Specification DEN0137 Sect. D2.1
[2] #964

GitHub-Depends-On: #972 
GitHub-Depends-On: #1212
GitHub-Depends-On: #1346

michpappas committed:
plat/common: Do not pass MEMRF_MAP when allocating memregs

UKPLAT_MEMRF_MAP / UKPLAT_MEMRF_UNMAP mrd types have been obsoleted by
the reworked implementation of paged memory init.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/kvm: Do not pass MEMRF_MAP when allocating memregs

UKPLAT_MEMRF_MAP / UKPLAT_MEMRF_UNMAP mrd types have been obsoleted by
the reworked implementation of paged memory init.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/pl011: Clean up driver init

Clean up pl011 init. Functional changes:
 - Return an error when init fails
 - Downgrade diagnostic message severity to uk_pr_debug

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/ns16550: Clean up driver init

Clean ns16550 driver init. Functional changes:
- Return an error when init fails
- Downgrade diagnostic message severity to uk_pr_debug

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/common: Add UKPLAT_MEMRT_DEVICE type

Regions of this type are added by device drivers that implement
an early init. Specifically, upon completion of the earlyinit boot
stage, device regions are expected to be mapped with appropriate
protections, and additionally be added to bootinfo using the
UKPLAT_MEMRT_DEVICE type.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/xen: Do not pass MEMRF_MAP when allocating memregs

UKPLAT_MEMRF_MAP / UKPLAT_MEMRF_UNMAP mrd types have been obsoleted by
the reworked implementation of paged memory init.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
include/memory: Retire UKPLAT_MEMRF_MAP / UKPLAT_MEMRF_UNMAP

UKPLAT_MEMRF_MAP / UKPLAT_MEMRF_UNMAP mrd types have been obsoleted by
the reworked implementation of paged memory init.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/ns16550: Fix config option of early console

Update the early console Kconfig option of ns16550 to aligned with the
current convention as well as the equivalent option of pl011.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/ns16550: Add early_init

Add early device init. When paging is enabled, this adds an mrd for
the ns16550 MMIO region to bootinfo so that the region is not unmapped
during paged memory init.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/kvm/arm: Add early init boot stage

Add boot stage for early initialization. This is invoked at the end
of early boot code, before passing control to the platform.

Early devices can use the APIs provided by the boot protocol to obtain
any information required, such as device regions and the kernel
command line.

Drivers that register with early_init() should append mrds of their
MMIO regions to bootinfo so that these regions are not unmapped
during paged memory init. These mrds must use the newly introduced
UKPLAT_MEMRT_DEVICE type.

Notice that early drivers should not call ukplat_bootinfo_coalesce(),
as mrd coalescing is performed once at the end of early_init().

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/kvm/arm64: Remove console registration from platform setup

With both UART drivers initialized by uk_inittab, remove console
registration from platform setup.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/pl011: Move driver initialization to initttab

Move console initialization with the rest of devices at
UK_INIT_CLASS_SYS. Since the pf bus does not support priority
levels, and the console should start before the pf bus to
allow drivers to print their status, register directly with
init instead of the pf bus.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/pl011: Add early init

Add early device init. When paging is enabled, this adds an mrd for
the pl011 MMIO region to bootinfo so that the region is not unmapped
during paged memory init.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/kvm/arm64: Register UART devices to early init

Register ns16550 and pl011 into ukplat_early_init.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/pl011: Map region on runtime

Map the pl011 region at runtime. This is now required as paged memory
init unmaps any memory not registered by early devices, thus if early
UART is not enabled the device regions is not mapped.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/ns16550: Map device region at runtime

Map the ns16550 region at runtime. This is now required as paged memory
init unmaps any memory not registered by early devices, thus if early
UART is not enabled the device regions is not mapped.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/ns16550: Move driver initialization to uk_inittab

Move console initialization with the rest of devices at
UK_INIT_CLASS_SYS. Since the pf bus does not support priority
levels, and the console should start before the pf bus to
allow drivers to print their status, register directly with
init instead of the pf bus.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
arch/arm64: Add definitions for block-size mappings

VMSAv8-64 does not provide a naming scheme for the block
size mapped by PT block descriptors at various translation
levels. Moreover, the block size varies depending on the
size of the translation granule.

To provide granularity agnostic definitions, use the
x86_64 terminology of Large / Huge pages.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/common: Rework paged memory init

Rework the initialization of paged memory to provide a more flexible
implementation that is capable of handling regions beyond the limits
defined in the boot pagetables. The motivation for this change is to
allow mapping device regions that are unknown at compile-time, such
as Unprotected IPA Alias regions of Arm CCA Realms, the address of
which depends on the executing platform.

Under the new scheme bootinfo is reduced to only contain mrds that
correspond to valid memory regions. This deprecates the unmap_mrd
region and the UKPLAT_MEMRF_MAP / UKPLAT_MEMRF_UNMAP mrd flags.
Moreover, the boot pagetables are no longer updated during paged
memory init, but instead are replaced with a new pagetable that
initialized with the regions defined in bootinfo. Besides the
additional flexibility, this implementation has the potential of
some performance improvement as it removes expensive TLB flush
operations associated with unmap.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>
Co-authored-by: Sergiu Moga <sergiu@unikraft.io>

michpappas committed:
plat/common/arm64: Set DIRECTMAP_AREA_END to the end of low VA range

The direct-mapped area maps the first 512GiB of the address space
to an architecture-defined region. In arm64 that uses the highest
512GiB of the low VA range. Update DIRECTMAP_AREA_END to correctly
specify the end of the low VA range.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukbus/platform: Update uk_bus_bf_devmap() to operate per-page

Update uk_bus_pf_devmap() to handle a multi-page region page-by-page,
to avoid an error caused by a partially mapped device region, which
would cause ukplat_page_map() would return EEXIST and in turn cause
the subsequent ukplat_page_set_attr() to fail.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/common: Move vaddr check to callers of pgarch_page_mapx()

Move vaddr check from pgarch_page_mapx() to its callers, as that
function is also used to map the direct-mapped region, the vaddr
of which is past (__VADDR_MAX - len).

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
As pointed out by @marcrittinghaus during a private discussion, the proposed way of mapping regions has a potential flaw if a region is partially mapped, as that would cause `ukplat_page_map()` to return `EEXIST`, which will in turn cause the subsequent `ukplat_page_set_attr()` to fail.

The implementation of paged memory init is now reworked to initialize a new page table instead (courtesy of @mogasergiu). Besides being simpler, this approach additionally avoids the expensive TLB invalidations associated with unmap.

mogasergiu commented on drivers/ukbus/platform/platform_bus.c:
```diff
@@ -259,17 +260,25 @@ __vaddr_t uk_bus_pf_devmap(__paddr_t base, __sz size)
+		if (!rc)
+			continue;
+
+		if (rc == -EEXIST)
```
I think we should keep the before behavior and crash instead. Something must be inconsistent in the memory map if we end up with an overlap, right?

michpappas replied:
So this is a fix regarding correctness as described in the commit message. Of course in pracrtice if that actually happens on device regions something has gone very wrong but even then I would leave handling to the caller instead. Either way that function is temporary, once we rework ukboot we should use be replacing this with `uk_vma_dma_map()`.

mogasergiu commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -65,29 +65,27 @@
 #define PL011_REG_READ(r)	ioreg_read16(PL011_REG(r))
 #define PL011_REG_WRITE(r, v)	ioreg_write16(PL011_REG(r), v)
 
-static void init_pl011(__u64 bas)
```
`bas`ed :moyai:

SerbanSo reviewed (commented):
The changes for the MRDs affecting kvm-x86 look good from my point of view.
However, I don't have enough code context for the changes to the drivers and the other architectures.

michpappas committed:
plat/common: Move coalesce out of bootinfo fdt setup

Remove coalescing out of bootinfo fdt setup to allow coalescing happen
at a boot protocol agnostic way at early_init(), after the initialization
of early devices.

Break down bootinfo_fdt_setup() into a pre and post coalesce functions
as the latter call ukplat_memory_alloc() which operates on ordered
regions.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/kvm: Move coalesce outside EFI bootinfo setup

Remove coalescing from EFI bootinfo setup to allow coalescing to
happen at a boot protocol agnostic way at early_init(), after the
initialization of early devices.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukintctlr/gic: Map GIC regions at runtime

Map GIC regions dynamically if paging is enabled. This is now required
as paged memory init unmaps any memory not backed by an mrd.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukrtc/pl031: Map pl031 regions at runtime

Map pl031 regions dynamically if paging is enabled. This is now
required as paged memory init unmaps any memory not backed by an
mrd.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/kvm/x86: Coalesce bootinfo at EFI post

With bootinfo coalesce having moved out of EFI common code,
and with kvm/x86 lacking an early init bootstage, do the coalescing
as the last part at EFI post before jumping to kernel.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/kvm: Clean up function declarations in EFI post

Remove redundant extern keyword from function declarations in EFI
post on arm64 and x86_64.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
Rebased from `staging`, no functional changes.

mogasergiu commented on plat/common/memory.c:
```diff
@@ -64,17 +68,12 @@ struct uk_alloc *ukplat_memallocator_get(void)
 	struct ukplat_memregion_desc *mrd, alloc_mrd = {0};
-	__vaddr_t unmap_start, unmap_end;
-	__sz unmap_len, desired_sz;
+	__sz desired_sz;
```
I thought you liked reverse xmas :(

michpappas replied:
Ooops thanks

mogasergiu commented on plat/kvm/arm/early_init.c:
```diff
@@ -0,0 +1,25 @@
+{
+	struct ukplat_bootinfo *bi;
+
+	bi = ukplat_bootinfo_get();
```
Why no `bi` check here?

mogasergiu commented on plat/kvm/arm/setup.c:
```diff
@@ -155,13 +155,13 @@ static void __noreturn _ukplat_entry2(void)
 
-	fdt = (void *)bi->dtb;
+	bi = ukplat_bootinfo_get();
 
```
Ditto?

mogasergiu commented on lib/ukbus/include/uk/bus.h:
```diff
@@ -42,9 +42,13 @@
-#define UK_BUS_INIT_CLASS UK_INIT_CLASS_PLAT
-#define UK_BUS_INIT_PRIO  0
-#define UK_BUS_REGISTER_PRIO 0
+#define UK_BUS_INIT_CLASS UK_INIT_CLASS_SYS
```
Do we know why it had PLAT prio before?

michpappas replied:
Good point, I assumed that this was since the platform bus was introduced, but now you mention I had a closer look and see that the change was added in 3092e3ccd1a62406af2bd49d47496a0f9e2581de with some justification on why it cannot be `CLASS_EARLY`, but not much justification on why it's set to `CLASS_PLAT`. My understanding that devices are initialized at `CLASS_SYS` is because of the relevant comment in https://github.com/unikraft/unikraft/blob/staging/include/uk/init.h#L125:
>  Define a stage for device initialization

HOWEVER: I now notice that `CLASS_SYS` states https://github.com/unikraft/unikraft/blob/staging/include/uk/init.h#L104:

> Define a stage for platform initialization. Platform at this point read
> all the device and device are initialized.

What I propose is that I revert this change and move also the initialization of late UART from SYS to PLAT. And let's clarify the scope of these two init levels with @skuenzer when we re-arch ukboot.

Thanks for catching that!

mogasergiu commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -105,39 +103,63 @@ static void init_pl011(__u64 bas)
 }
 
-void pl011_console_init(const void *dtb)
+int pl011_console_init(void *dtb)
```
Why is the `const` removed?

michpappas replied:
Honestly I don't remember, but the interface of this function is updated on a later commit anyway. If you think it's necessary I can revert this change though.

mogasergiu replied:
Meh, if it's overwritten anyway I would say don't bother :shrimp: :fried_shrimp: :man_shrugging:

mogasergiu commented on drivers/uktty/pl011/exportsyms.uk:
```diff
@@ -1,3 +1,4 @@
+pl011_early_init
```
Commit message of
> drivers/uktty/pl011: Add early init

Add early device init. When paging is enabled, this adds an mrd for
the pl011 MMIO region to bootinfo so that the region is **mapped after paged memory init**.

michpappas replied:
Currently the commit message states:
> region is not unmapped during paged memory init.

That is actually accurate. MMIO regions of early devices are already mapped, and the purpose of adding an mrd to bootinfo is so that the device does not get unmapped during paged memory init (i.e. in the mapping makes it to the new pagetable and the region stays mapped).

mogasergiu commented on drivers/uktty/ns16550/ns16550.c:
```diff
@@ -126,35 +123,49 @@ static void init_ns16550(__u64 base)
 }
 
-void ns16550_console_init(const void *dtb)
+int ns16550_console_init(void *dtb)
```
Ditto const?

michpappas replied:
See https://github.com/unikraft/unikraft/pull/1373#discussion_r1616815507

mogasergiu commented on drivers/ukintctlr/gic/gic-v2.c:
```diff
@@ -608,6 +613,29 @@ static int gicv2_do_probe(void)
+	vbase = uk_bus_pf_devmap(gicv2_drv.dist_mem_addr,
+				 gicv2_drv.dist_mem_size);
+	if (unlikely(PTRISERR(vbase)))
+		return PTR2ERR(vbase);
```
No error messages?

mogasergiu commented on drivers/ukintctlr/gic/gic-v2.c:
```diff
@@ -641,6 +669,12 @@ int gicv2_probe(struct _gic_dev **dev)
+#if CONFIG_PAGING
+	rc = gicv2_map();
+	if (unlikely(rc))
+		return rc;
```
Ditto

mogasergiu commented on drivers/ukintctlr/gic/gic-v3.c:
```diff
@@ -692,6 +697,29 @@ static int gicv3_do_probe(void)
 #endif /* !CONFIG_UKPLAT_ACPI */
 
+#if CONFIG_PAGING
+static int gicv3_map(void)
```
Ditto for GICv3

mogasergiu commented on drivers/ukrtc/pl031/pl031.c:
```diff
@@ -180,6 +184,12 @@ int pl031_init_rtc(void *dtb)
+#if CONFIG_PAGING
+	pl031_base_addr = uk_bus_pf_devmap(pl031_base_addr, size);
+	if (unlikely(PTRISERR(pl031_base_addr)))
+		return PTR2ERR(pl031_base_addr);
```
Ditto?

mogasergiu commented on drivers/uktty/ns16550/ns16550.c:
```diff
@@ -126,35 +128,100 @@ static void init_ns16550(__u64 base)
+	ns16550_uart_initialized = 1;
+
+	bi = ukplat_bootinfo_get();
+	if (unlikely(!bi)) {
```
Hmm, should probably just have an assertion here, as the getter already checks stuff.

michpappas replied:
~~Generally speaking I'd rather not make assumptions on a function's implementation and rather rely on the interface, given that if the implementation changes all calling code breaks. Besides, from a quick look I can see several places where we check the result of `ukplat_bootinfo_get()` rather than asserting.~~

Actually nevermind, although the above is conceptually correct an assertion seems more appropriate in this context 👍🏼

mogasergiu replied:
I feel like the problem here may also lie in the fact that it's kinda impossible for this getter to fail (or is it?). Maybe a dtb getter should exist and be used instead. Anyway no matter, too much of a trivial thing that will get overwritten by rearch/refactor anyway. I will let you choose whatever feels more sane here.

mogasergiu commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -105,39 +104,109 @@ static void init_pl011(__u64 bas)
+int pl011_early_init(void)
+{
+#if CONFIG_LIBUKTTY_PL011_EARLY_CONSOLE
+	struct ukplat_bootinfo *bi = ukplat_bootinfo_get();
```
assert `bi`?

mogasergiu commented on plat/common/bootinfo_fdt.c:
```diff
@@ -198,14 +198,22 @@ void ukplat_bootinfo_fdt_setup(void *fdtp)
+	struct ukplat_bootinfo *bi;
+
+	bi = ukplat_bootinfo_get();
+	if (unlikely(!bi))
```
Ditto

michpappas replied:
I follow the pattern used in the rest of this file by the original author:
```
1afb423425 (Sergiu Moga     2023-05-14 18:21:55 +0300 187) void ukplat_bootinfo_fdt_setup(void *fdtp)
1afb423425 (Sergiu Moga     2023-05-14 18:21:55 +0300 188) {
1afb423425 (Sergiu Moga     2023-05-14 18:21:55 +0300 189)      struct ukplat_bootinfo *bi;
1afb423425 (Sergiu Moga     2023-05-14 18:21:55 +0300 190) 
1afb423425 (Sergiu Moga     2023-05-14 18:21:55 +0300 191)      bi = ukplat_bootinfo_get();
1afb423425 (Sergiu Moga     2023-05-14 18:21:55 +0300 192)      if (unlikely(!bi))
1afb423425 (Sergiu Moga     2023-05-14 18:21:55 +0300 193)              ukplat_bootinfo_crash("Invalid bootinfo");
```

mogasergiu replied:
The **unbelievably handsome and charismatic** author may have just changed his mind :thinking:, I am joking (about the latter). In all seriousness though, I believe these changes were back when I barely had any familiarity with this code I guess. I am fine with leaving as is and coming back when we rearch ukboot.

mogasergiu commented on plat/common/memory.c:
```diff
@@ -133,7 +120,7 @@ void *ukplat_memregion_alloc(__sz size, int type, __u16 flags)
-		alloc_mrd.flags = flags | UKPLAT_MEMRF_MAP;
+		alloc_mrd.flags = flags;
 
 		bi = ukplat_bootinfo_get();
```
I know I wrote this :laughing: but ditto, this is a crash reason tbh.

michpappas replied:
In principle I agree, but taking into account earlier comments on the same topic, and the number places in the code base where we actually check the return value of `ukplat_bootinfo_get()`, I would rather if we came up with a common approach that is also expressed on the interface of this function, rather than ad-hoc requests on reviews. Either way, I would leave this as is given that it was not changed in this PR and the error is propagated to the caller anyway.

mogasergiu replied:
> The **unbelievably handsome and charismatic** author may have just changed his mind 🤔, I am joking (about the latter). In all seriousness though, I believe these changes were back when I barely had any familiarity with this code I guess. I am fine with leaving as is and coming back when we rearch ukboot.

Ditto. I am fine with leaving for later.

mogasergiu reviewed (changes requested):
Hey, LGTM :+1:. Couple of comments. I will then proceed to have an additional look and test on x86 UEFI. Thanks.

michpappas commented:
@mogasergiu I addressed all review comments with a thumbs-up.

For the `ukplat_bootinfo_get()` I agree in principle, but in that case we should consider updating the interface to reflect that this function should always succeed and as such any failures should be caught via assertions.

I also tested the latest state with:
* `arm64/qemu`: `app-helloworld` with `lxboot`, `virt`, and `UEFI` OK ✅
* `arm64/fircracker`: ~~compile only, may fire a baremetal instance for a runtime test later~~ OK ✅
* `x86_64/firecracker` with `app-helloworld` OK ✅
* `x86_64/qemu`: `app-helloworld` with lxboot OK ✅
* `x86_64/qemu`: `app-helloworld` with EFI gives me an error: `qemu-system-x86_64: combined size of system firmware exceeds 8388608 bytes` but I'm not very familiar with `qemy-system-x86_64` so it may be that I'm doing something wrong.

mogasergiu reviewed (approved):
Me gusta. Tested on x86 UEFI as well.

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

michpappas commented:
Pushed previously missed revert for https://github.com/unikraft/unikraft/pull/1373#discussion_r1616084891.

michpappas commented:
Rebased from `unikraft/staging`

SerbanSo reviewed (approved):
Looks good to me.

Reviewed-by: Serban Sorohan <serban.sorohan@gmail.com>

michpappas commented:
Rebased from `staging` and resolved conflict with #1434

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
